library/openssl/kdf/pbkdf2_hmac_spec.rb

```diff
@@ -0,0 +1,283 @@
+    length = 16
+    hash = "wd40"
+    -> {
+      OpenSSL::KDF.pbkdf2_hmac(pass, salt: salt, iterations: iterations, length: length, hash: hash)
+    }.should raise_error(RuntimeError, /Unsupported digest algorithm \(wd40\)/)
+  end
+
+  it "treats salt as a required keyword" do
+    pass = "secret"
+    iterations = 20_000
+    length = 16
+    hash = "sha1"
+    -> {
+      OpenSSL::KDF.pbkdf2_hmac(pass, iterations: iterations, length: length, hash: hash)
+    }.should raise_error(ArgumentError, 'missing keyword: :salt')
+  end
+
+  it "treats iterations as a required keyword" do
+    pass = "secret"
+    salt = "\x00".b * 16
+    length = 16
+    hash = "sha1"
+    -> {
+      OpenSSL::KDF.pbkdf2_hmac(pass, salt: salt, length: length, hash: hash)
+    }.should raise_error(ArgumentError, 'missing keyword: :iterations')
+  end
+
+  it "treats length as a required keyword" do
+    pass = "secret"
+    salt = "\x00".b * 16
+    iterations = 20_000
+    hash = "sha1"
+    -> {
+      OpenSSL::KDF.pbkdf2_hmac(pass, salt: salt, iterations: iterations, hash: hash)
+    }.should raise_error(ArgumentError, 'missing keyword: :length')
+  end
+
+  it "treats hash as a required keyword" do
+    pass = "secret"
+    salt = "\x00".b * 16
+    iterations = 20_000
+    length = 16
+    -> {
+      OpenSSL::KDF.pbkdf2_hmac(pass, salt: salt, iterations: iterations, length: length)
+    }.should raise_error(ArgumentError, 'missing keyword: :hash')
+  end
+
+  it "treats all keywords as required" do
+    pass = "secret"
+    -> {
+      OpenSSL::KDF.pbkdf2_hmac(pass)
+    }.should raise_error(ArgumentError, 'missing keywords: :salt, :iterations, :length, :hash')
+  end
+
+
+  it "has behaviour for 0 or less iterations that depends on the version of OpenSSL" do
+    pass = "secret"
+    salt = "\x00".b * 16
+    length = 16
+    hash = "sha1"
+    if OpenSSL::VERSION.split('.').first.to_i < 3
+      # "Any iter less than 1 is treated as a single iteration."
+      iterations = 0
+      key0 = OpenSSL::KDF.pbkdf2_hmac(pass, salt: salt, iterations: iterations, length: length, hash: hash)
+      iterations = -1
+      key_negative = OpenSSL::KDF.pbkdf2_hmac(pass, salt: salt, iterations: iterations, length: length, hash: hash)
+      iterations = 1
+      key1 = OpenSSL::KDF.pbkdf2_hmac(pass, salt: salt, iterations: iterations, length: length, hash: hash)
+      key0.should == key1
+      key_negative.should == key1
+    else
+      # This raises an error (even though the OpenSSL docs say otherwise)
+      iterations = 0
+      -> {
+        OpenSSL::KDF.pbkdf2_hmac(pass, salt: salt, iterations: iterations, length: length, hash: hash)
+      }.should raise_error(OpenSSL::KDF::KDFError, "PKCS5_PBKDF2_HMAC: invalid iteration count")
+
+      iterations = -1
+      -> {
+        OpenSSL::KDF.pbkdf2_hmac(pass, salt: salt, iterations: iterations, length: length, hash: hash)
+      }.should raise_error(OpenSSL::KDF::KDFError, /PKCS5_PBKDF2_HMAC/)
+    end
```

Comment:
**suggestion:** there is a conventional way to express different behaviour depending on platform/OS/library version - with `guard` DSL helper, e.g.:

```ruby
  guard -> { Psych::VERSION < "4.0.0" } do
    it_behaves_like :yaml_load_unsafe, :load
  end
```

Reply:
That seems to work, tested it on an OpenSSL 1.1 VM that I still had somewhere.